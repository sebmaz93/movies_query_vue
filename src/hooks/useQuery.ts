import { ref } from 'vue';

// types
import type { AxiosResponse } from 'axios';

export const useQuery = <T, F>(
  key: string,
  fn: (q?: F) => Promise<AxiosResponse<T>>,
) => {
  const data = ref<T>();
  const error = ref<string>();
  const isLoading = ref(false);

  const handleError = (error: any) => {
    isLoading.value = false;
    error.value = error.message;
    console.log(`[NETWORK:${key}:ERR]:`, error);
  };

  const fetch = async (query?: F) => {
    isLoading.value = true;
    console.log(`[NETWORK:${key}:REQ]:`, query);
    try {
      const response = await fn(query);
      data.value = response.data;
      console.log(`[NETWORK:${key}:RES]:`, response.data);
      isLoading.value = false;
      console.log('data', data);
    } catch (error) {
      handleError(error);
    }
  };

  return {
    data,
    error,
    isLoading,
    fetch,
  };
};
