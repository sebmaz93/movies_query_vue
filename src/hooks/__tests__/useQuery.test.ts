import { AxiosResponse } from 'axios';

import { useQuery } from '../useQuery';
import moviesApi from '@services/api/movie';

jest.mock('vue', () => {
  return {
    ...jest.requireActual('vue'),
    onMounted: jest.fn((cb) => cb()),
  };
});

describe('useQuery', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  it('should invoke the passed function when calling fetch', () => {
    const mockFn = jest
      .fn()
      .mockImplementation(
        () => new Promise<Partial<AxiosResponse>>((resolve) => resolve({ data: {} }))
      );

    const { fetch } = useQuery('fn', mockFn);

    fetch();

    expect(mockFn.mock.calls).toMatchSnapshot();
  });

  it('should fetch movies', () => {
    const spy = jest.spyOn(moviesApi, 'fetchMovies').mockImplementation(
      () =>
        new Promise((resolve) =>
          resolve({
            data: {
              page: 2,
              total_pages: 3,
              has_next_page: true,
              results: [],
              total_results: 14,
            },
          } as any)
        )
    );

    const { fetch } = useQuery('GET MOVIES', moviesApi.fetchMovies);

    fetch({ page: 2, per_page: 6 });

    expect(spy.mock.calls).toMatchSnapshot();
  });
});
