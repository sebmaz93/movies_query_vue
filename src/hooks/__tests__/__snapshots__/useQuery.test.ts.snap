// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useQuery should fetch movies 1`] = `
Array [
  Array [
    Object {
      "page": 2,
      "per_page": 6,
    },
  ],
]
`;

exports[`useQuery should invoke the passed function when calling fetch 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;
